### PR TITLE
[FEATURE] Instantiate Rule-Based Profiler Using Typed Configuration Object

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-564/GREAT-568
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: just wanted to check my understanding here: so the `RuleBasedProfiler` has no required parameters to instantiate?

Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: shouldn't we at least have a name?

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Thank you for this

**Comment by Shinnnyshinshin** [comment]
Thank you for this Alex, one question regarding the constructor

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @Shinnnyshinshin It does have a name -- it can be supplied explicitly, or as part of the `RuleBasedProfilerConfig` typed object (but not both).  Thanks.

Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: Ok that makes a lot of sense. Right now, I think the logic only tests for `profiler_config` not being None. Could we add additional checks (and associated error messages) for a situation where 1) both are provided 2) neither are provided?

Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: both meaning `profiler_config` and individual parameters

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @Shinnnyshinshin This should be done outside -- by running it through YAML/config checker.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @Shinnnyshinshin constructor full of nulls will raise an error because domain_builder is required in a rule.

Comment (donaldheppner) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: I see what is going on here; we are allowing both config-based initialization or a "decomposed config initialization". My preference would be to pick one (the config-based one), especially since the only thing calling this constructor is us, and make all the parameters required. Kind of a "why do more when you can do less?" approach.

It might be work revisiting this constructor when we remove DataContext as a dependency, but for now we can tolerate it, assuming there is a good reason to allow for both code paths/methods of initialization

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:616]: Does this need to be public?

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Couple of questions:

1. (most important) Why are setting and getting rules using different backing stores?
2. (more philosophical) How/who consumes the rules? Is there an opportunity to encapsulate that work within the rule-based profiler?

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:616]: @donaldheppner Yes -- for the upcoming work on the "Self-Initializing Expectations".  Thanks. P.S.: Please do not merge yet (I am following up on @Shinnnyshinshin 's feedback and will commit an update soon).

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @donaldheppner 1) Store is not present here as far as I can tell (or did you mean something else by "store"?). 2) The rules are consumed by profiler and also in the case of the self-initializing expectation, exactly one (or zero) rule will be needed (hence the public method).

Comment (donaldheppner) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: by store I meant different attributes; the "get" property reads from a dictionary and the "setter" initializes `_rules`

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @donaldheppner Yes, `Rules` are initialized (from config), but they never set in the standard "obj.prop = val" sense (we have not encountered the need for defining a setter, because no external object has required it so far).  There is a `@property` getter, because the rules do need to be obtained by other objects (and tule `RuleBasedProfiler` internal usage, too), and finally, there is a convenience method that returns rules (which are formally a list) as dictionary with the rule name as the key (again, this is a "helper" as a convenience for the needs of other methods).  Thanks.

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:616]: I have the same question - I'd love to connect on this synchronously if possible.

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:665]: Was this inheritance hierarchy always the plan? I like this but I'd love to think out to possible design implications of something like this.

If we're using a `Base` object here, we should make it a pattern and add it to our core concepts doc.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:665]: @cdkini Happy to discuss -- For now, the existing pattern, such as `BaseDataContext` is instantiated with the typed Config objects and subclasses with `**kwargs` is the only way our approach of materializing classes from configuration using `instantiate_class_from_config()` will work.

**Review by Shinnnyshinshin — APPROVED** [review]
Love the changes @alexsherstinsky. LGTM pending @donaldheppner and @cdkini questions being addressed

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:616]: @cdkini -- @donaldheppner and I discussed the need for reconciliation.  Also, I was able to make the `rules` setter "standard" -- please re-review.  Thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @cdkini -- @donaldheppner and I discussed the need for reconciliation. Also, I was able to make the rules setter "standard" -- please re-review. Thanks!

**Review by cdkini — APPROVED** [review]

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**
